packages/video_player/video_player/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java

```diff
@@ -50,27 +51,51 @@
 
   private final TextureRegistry.SurfaceTextureEntry textureEntry;
 
-  private QueuingEventSink eventSink = new QueuingEventSink();
+  @VisibleForTesting QueuingEventSink eventSink = new QueuingEventSink();
 
   private final EventChannel eventChannel;
 
-  private boolean isInitialized = false;
+  @VisibleForTesting boolean isInitialized = false;
 
   private final VideoPlayerOptions options;
 
+  VideoPlayer(
+          Context context,
+          EventChannel eventChannel,
+          TextureRegistry.SurfaceTextureEntry textureEntry,
+          String dataSource,
+          String formatHint,
+          Map<String, String> httpHeaders,
+          VideoPlayerOptions options) {
+    this(
+            context,
+            eventChannel,
+            textureEntry,
+            dataSource,
+            formatHint,
+            httpHeaders,
+            options,
+            null);
+  }
+
   VideoPlayer(
       Context context,
       EventChannel eventChannel,
       TextureRegistry.SurfaceTextureEntry textureEntry,
       String dataSource,
       String formatHint,
       Map<String, String> httpHeaders,
-      VideoPlayerOptions options) {
+      VideoPlayerOptions options,
+      SimpleExoPlayer exoPlayer) {
     this.eventChannel = eventChannel;
     this.textureEntry = textureEntry;
     this.options = options;
 
-    exoPlayer = new SimpleExoPlayer.Builder(context).build();
+    if (exoPlayer != null) {  // Skip remaining setup in tests if using a mock exoPlayer.
```

Comment:
I think the logic would be much clearer here (and less error-prone; someone could easily add non-ExoPlayer-specific code after this block on accident) to have the other constructor do ExoPlayer setup and pass in the ExoPlayer, instead of passing null and then handling null specially here.

(Then you can also drop some parameters from this constructor, which will make it less confusing, since currently it sometimes ignores a bunch of parameters.)

Reply:
I'll leave the original constructor alone and just create a VisibleForTesting constructor for setting required (private final) fields to mocks.

(I like your idea of setting up the ExoPlayer then passing it in, but after setup there's really no remaining logic for a secondary constructor to handle.)

Reply:
(Done in https://github.com/flutter/plugins/pull/3820/commits/5f95287a54b82ca90cde6153285ab9e9b10e2493)